scripts/emacs_lua_complete.lua

```diff
@@ -0,0 +1,79 @@
+--    __emacs_lua_complete:
+-- This function finds and prints a limited number of matches on the
+-- lua globals tree, _G. See `lua-complete-string`. N.B.: Does not
+-- complete numeric keys (e.g. for arrays) or keys which begin with an
+-- underscore.
+-- Arguments:
+--   PARTS: a table array of dot-separated parts of the completion item
+--     (e.g. table.con would become {"table","con"}).  All but the last
+--     element must be valid subkeys in the _G heirarchy.
+--   LIBS: an array of tables of the form
+--       {var: varname, lib: libfile}
+--     for e.g. local varname=require("libfile"). This library will be
+--     require'd and searched alongside global variables.
+--   LOCALS: an array of strings parsed from local aloc,bloc = ... to
+--     complete as well
+--
+-- Example:
+-- __emacs_lua_complete({'table','ins'},{{var="xyz",lib="libfile"}},
+--                      {"frank","beans"},5000)
+--
+-- Written 2019/05 J.D. Smith
+function __emacs_lua_complete(parts, libs, locals, limit)
+   -- Setup a FIFO queue for breadth-first tree traversal
+   local queue = {head = 0, tail = 0}
+   function queue:push(value)
+      self.tail = self.tail+1  self[self.tail] = value
+   end
+   function queue:pop()
+      self.head = self.head+1  return self[self.head]
+   end
+
+   -- Mirror the globals table
+   local globals,cnt,limit = {}, 0, limit or 2000
+   for k,v in pairs(_G) do globals[k] = v end
+   
+   -- Add libs and locals
+   if libs then
+      for _,v in ipairs(libs) do
+	 local good,lib=pcall(require,v.lib)
```

Comment:
When trying out stuff locally I have realized that this probably doesn't reload the changes that were made after the first completion attempt. Should we force a reload here, WDYT?

Reply:
It's only if the last completion was a substring of the current completion string (modulo whitespace) that the caching is used. So I don't see how you could get into trouble.  Maybe if you reuse the same local variable but load a different library that might be problematic?  Maybe we should make an explicit binding to drop the last saved completion by setting `lua-last-completed` to nil?

Reply:
I'm not referring to the `lua-last-completed` caching, I'm talking about the "caching" mechanism behind Lua's `require` function. IIRC it won't reload a library if you `require` it for the second time from the same interpreter.

Reply:
Oh I see.  I thought `require` checked for changes on disk?  Would `dofile` work if not?

Reply:
I think it should be easier to consult `require.loaded` table and remove an entry from there than reimplementing `require`'s path lookups if I'm reading [the doc](http://www.lua.org/manual/5.1/manual.html#pdf-require) correctly.

Reply:
This seems to me to be beyond the scope of `lua-mode`, since people need to come up with some solution for reloading modules which are changing on disk anyway, even for just _running code which uses the module_.  

I'm suprised `require` doesn't offer changed file reloading.  But since it doesn't, most people would reasonably expect that until they actively "re-require" the module (by clearing it from `package.loaded`, for example), it would remain unchanged.  If `lua-mode` silently re-requires modules in the background through the course of completion, that would invalidate that likely assumption, and result in confusion.  It could also be a big performance hit for large modules to constantly reload them.  

I'm inclined to leave this one to the user to solve (as they already must!).  Or are we looking at it differently?

Reply:
Ok, let's leave this for posterity. You are right that live reloading modules as a side-effect of completions in the subshell should not happen as it is very surprising.

But in general, that's the thing with completion, the user doesn't really care about how the completions are done. I can say for myself, if I change a module in my project I would expect the new changes to be listed in completions right away. I don't care where do they come from, are they cached or are they reloaded on every request (provided it's fast enough). Probably piggybacking the completion functionality on the running Lua inferior shell should be rethought in the future.

Reply:
OK, thanks.  One future solution is we could provide a convenient shortcut for "dump the `package.loaded` of all the modules `require`d in this file".  That would then be of general use for better completion and inheriting new Lua behavior on running your code. And yes, parsing the Lua code directly is more flexible (and way more complicated, I know from experience with another major mode of mine for the IDL language).